clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2600,29 +2646,230 @@ namespace {
+                                          CastKind::CK_ArrayToPointerDecay, E);
+          return ExpandToRange(Base, B);
+        }
+
+        // If B is an interop type annotation, the type must be identical
+        // to the declared type, modulo checkedness.  So it is OK to
+        // compute the array bounds based on the original type.
+        return ArrayExprBounds(E);
+      }
+
+      // Infer the target bounds of e.
+      // e only has target bounds if e does not have array type.
+      bool IsParam = isa<ParmVarDecl>(E->getDecl());
+      if (E->getType()->isCheckedPointerPtrType())
+        OutTargetBounds = CreateTypeBasedBounds(E, E->getType(),
+                                                IsParam, false);
+      else if (!VD)
+        OutTargetBounds = CreateBoundsInferenceError();
+      else if (!B && IT)
+        OutTargetBounds = CreateTypeBasedBounds(E, IT->getType(),
+                                                IsParam, true);
+      else if (!B || B->isUnknown())
+        OutTargetBounds = CreateBoundsAlwaysUnknown();
+      else {
+        Expr *Base = CreateImplicitCast(E->getType(),
+                                        CastKind::CK_LValueToRValue, E);
+        OutTargetBounds = ExpandToRange(Base, B);
+      }
+
+      if (E->getType()->isFunctionType()) {
+        // Only function decl refs should have function type
+        assert(isa<FunctionDecl>(E->getDecl()));
+        return CreateBoundsEmpty();
+      }
+
+      Expr *AddrOf = CreateAddressOfOperator(E);
+      return CreateSingleElementBounds(AddrOf);
+    }
+
+    // If e is an lvalue, CheckUnaryLValue returns the
+    // lvalue and target bounds of e.
+    // If e is an rvalue, CheckUnaryOperator should be called instead.
+    BoundsExpr *CheckUnaryLValue(UnaryOperator *E,
+                                 BoundsExpr *&OutTargetBounds) {
+      BoundsExpr *SubExprBounds = Check(E->getSubExpr());
+
+      if (E->getOpcode() == UnaryOperatorKind::UO_Deref) {
+        // Currently, we don't know the target bounds of a pointer stored in a
+        // pointer dereference, unless it is a _Ptr type (handled
+        // earlier) or an _Nt_array_ptr.
+        if (E->getType()->isCheckedPointerNtArrayType())
+          OutTargetBounds = CreateTypeBasedBounds(E, E->getType(),
+                                                  false, false);
+        else
+          OutTargetBounds = CreateBoundsUnknown();
+
+        // The lvalue bounds of *e are the rvalue bounds of e.
+        return SubExprBounds;
+      }
+
+      OutTargetBounds = CreateBoundsInferenceError();
+      return CreateBoundsInferenceError();
+    }
+
+    // CheckArraySubscriptExpr returns the lvalue and target bounds of e.
+    // e is an lvalue.
+    BoundsExpr *CheckArraySubscriptExpr(ArraySubscriptExpr *E,
+                                        BoundsExpr *&OutTargetBounds) {
+      // Currently, we don't know the target bounds of a pointer returned

```

Comment:
It isn't clear that _Ptr is handled earlier.

Reply:
This was also left over from the top-down approach and has been updated.